vms/avm/states/state.go

```diff
@@ -576,3 +597,175 @@ func (s *state) writeStatuses() error {
 	}
 	return nil
 }
+
+func (s *state) Prune(lock sync.Locker, log logging.Logger) error {
+	lock.Lock()
+	// It is possible that more txs are added after grabbing this iterator. No
+	// new txs will write a status, so we don't need to check those txs.
+	statusIter := s.statusDB.NewIterator()
+	defer statusIter.Release()
+
+	if !statusIter.Next() {
+		// If there are no statuses on disk, pruning was previously run and
+		// finished.
+		lock.Unlock()
+
+		log.Info("state already pruned")
+
+		return statusIter.Error()
+	}
+
+	startTxIDBytes := statusIter.Key()
+	txIter := s.txDB.NewIteratorWithStart(startTxIDBytes)
+	defer txIter.Release()
+
+	// While we are pruning the disk, we disable caching of the data we are
+	// modifying. Caching is re-enabled when pruning finishes.
+	//
+	// Note: If an unexpected error occurs the caches may never be re-enabled.
+	// That's fine as the node is going to be in an unhealthy state regardless.
+	oldStatusCache := s.statusCache
+	oldTxCache := s.txCache
+	s.statusCache = &cache.Empty[ids.ID, *choices.Status]{}
+	s.txCache = &cache.Empty[ids.ID, *txs.Tx]{}
+	lock.Unlock()
+
+	startTime := time.Now()
+	startProgress := timer.ProgressFromHash(startTxIDBytes)
+
+	startStatusBytes := statusIter.Value()
+	if err := s.cleanupTx(lock, startTxIDBytes, startStatusBytes, txIter); err != nil {
+		return err
+	}
+
+	i := 1
+	for statusIter.Next() {
+		i++
+
+		txIDBytes := statusIter.Key()
+		statusBytes := statusIter.Value()
+		if err := s.cleanupTx(lock, txIDBytes, statusBytes, txIter); err != nil {
+			return err
+		}
+		if i%pruneCommitLimit == 0 {
+			// We must hold the lock during committing to make sure we don't
+			// attempt to commit to disk while a block is concurrently being
+			// accepted.
+			lock.Lock()
+			err := s.Commit()
+			lock.Unlock()
+			if err != nil {
+				return err
+			}
+
+			progress := timer.ProgressFromHash(txIDBytes)
+			eta := timer.EstimateETA(
+				startTime,
+				progress-startProgress,
+				math.MaxUint64-startProgress,
+			)
+			log.Info("committing state pruning",
+				zap.Int("count", i),
+				zap.Duration("eta", eta),
+			)
+		}
+	}
+
+	lock.Lock()
+	defer lock.Unlock()
+
+	errs := wrappers.Errs{}
+	errs.Add(
+		s.Commit(),
+		statusIter.Error(),
+		txIter.Error(),
+	)
+
+	// Make sure we flush the original caches before re-enabling them to prevent
+	// surfacing any stale data.
+	oldStatusCache.Flush()
+	oldTxCache.Flush()
+	s.statusCache = oldStatusCache
```

Comment:
Why not disable db status lookup when complete (and disable this cache)?

Reply:
Like always return `ErrNotFound` without performing any DB lookups?

We could... I just wanted to minimize the diff (we'll remove writing of statuses and reading of statuses with the removal of the `uniqueTx` struct.

Reply:
Yeah, that was my idea.